interface.h

```diff
@@ -64,150 +67,853 @@
+ */
 void set_er301_cv(uint8_t output, int16_t value);
+
+/**
+ * @brief Set the value of the indicated CV gate output on a connected ER301 via
+ *        I2C.
+ * 
+ * @param output ER301 CV gate output number. A valid value is between 0 and
+ *        MAX_ER301_OUTPUT_COUNT - 1
+ * @param on 0 CV gate output is low,
+ *           1 CV gate output is high
+ */
 void set_er301_gate(uint8_t output, uint8_t on);
 
+/**
+ * @brief Set the mode on a connected Just Friends module via I2C. Used to
+ *        switch Just Friends between a default behaviour and an alternate
+ *        operating mode that provides access to it's alternate personalities of
+ *        Synthesis, a polyphonic synthesizer (while in mode 1, and JF is set to
+ *        "sound"), and Geode a rhythm machine (while in mode 1, and JF is set
+ *        to "shape").
+ * 
+ * @param mode 0 default behaviour
+ *             1 activates alternate modes with any non-zero value treated as 1
+ */
 void set_jf_mode(uint8_t mode);
+
+/**
+ * @brief Set the value of the indicated CV gate output on a connected Just
+ *        Friends via I2C.
+ * 
+ * @param output Just Friends CV gate output number. A valid value is between 0
+ *        and MAX_JF_VOICE_COUNT - 1
+ * @param on 0 CV gate output is low,
+ *           1 CV gate output is high
+ */
 void set_jf_gate(uint8_t output, uint8_t on);
 
+/**
+ * @brief Set the envelope mode of the indicated output on a connected TELEXo
+ *        module via I2C.
+ * 
+ * @param output TELEXo CV output number. A valid value is between 0 and
+ *        MAX_TXO_VOICE_COUNT - 1
+ * @param mode TELEXo CV output envelope mode
+ *             0 sets output to oscillator mode and initializes the output to
+ *             waveform 0 (sine)
+ *             1 sets output to envelope mode
+ * @see set_txo_waveform()
+ */
 void set_txo_mode(uint8_t output, uint8_t mode);
+
+/**
+ * @brief Set the CV value of the indicated output on a connected TELEXo module
+ *        via I2C.
+ * 
+ * @param output TELEXo CV output number. A valid value is between 0 and
+ *        MAX_TXO_VOICE_COUNT - 1
+ * @param value CV value represented as a signed integer with a range of
+ *        -16,384 (-10v) and 16,383 (+10v)
+ */
 void set_txo_cv(uint8_t output, int16_t value);
+
+/**
+ * @brief Set the CV gate value of the indicated output on a connected TELEXo
+ *        module via I2C.
+ * 
+ * @param output TELEXo CV gate output number. A valid value is between 0 and
+ *        MAX_TXO_VOICE_COUNT - 1
+ * @param on 0 CV gate output is low,
+ *           1 CV gate output is high
+ */
 void set_txo_gate(uint8_t output, uint8_t on);
 
+/**
+ * @brief Set the envelope attack value of the indicated output on a connected
+ *        TELEXo module via I2C.
+ * 
+ * @param output TELEXo output number. A valid value is between 0 and
+ *        MAX_TXO_VOICE_COUNT - 1
+ * @param attack Output envelope attack rate in milliseconds. A valid value is
+ *        between 1 and n
```

Comment:
not actually sure what's the maximum value is - i think it takes a 16 bit parameter

Reply:
Looks like `UINT16_MAX` should work... https://github.com/bpcmusic/telex/blob/master/software/TELEXo/CVOutput.cpp#L517

Reply:
do you mean this line? `_attack = TxHelper::ConvertMs(max(att, 1), format);`

i think this is just meant to make 1 the smallest value.

Reply:
Totally.